File: backend/app/models/gain_passif.py
from ..extensions import db


class GainPassif(db.Model):
    """
    Gain ou perte passive par tour planifié (mercredi et samedi à minuit).
    Plusieurs entrées par (joueur, ressource) possibles.
    tours_restants NULL = définitif ; sinon nombre de tours restant avant désactivation.
    """

    __tablename__ = "gain_passif"

    id = db.Column(db.Integer, primary_key=True)
    utilisateur_id = db.Column(
        db.String(20), db.ForeignKey("utilisateur.id"), nullable=False, index=True
    )
    ressource_id = db.Column(
        db.Integer, db.ForeignKey("ressource.id"), nullable=False, index=True
    )
    quantite_par_tour = db.Column(db.Integer, nullable=False)
    actif = db.Column(db.Boolean, default=True, nullable=False)
    tours_restants = db.Column(db.Integer, nullable=True)
    # delai_tours : nombre de tours avant le démarrage effectif de la production
    # (ex : une "scierie" qui produit dans 2 tours => delai_tours=2).
    delai_tours = db.Column(db.Integer, nullable=False, default=0)
    # Optionnel : ligne créée par un évènement MJ (non éditable par un joueur)
    evenement_id = db.Column(db.Integer, db.ForeignKey("evenement.id"), nullable=True, index=True)
    # science | politique | evenement | batiment | autre
    balise = db.Column(db.String(20), nullable=False, default="autre")
    # fixe = unités ; pourcentage = % de la production cumulée du tour avant cette ligne (ordre id)
    mode_production = db.Column(db.String(20), nullable=False, default="fixe")

    ressource = db.relationship("Ressource")

    def to_dict(self):
        return {
            "id": self.id,
            "utilisateur_id": self.utilisateur_id,
            "ressource_id": self.ressource_id,
            "ressource": self.ressource.to_dict(),
            "quantite_par_tour": self.quantite_par_tour,
            "actif": self.actif,
            "definitif": self.tours_restants is None,
            "tours_restants": self.tours_restants,
            "delai_tours": self.delai_tours,
            "evenement_id": getattr(self, "evenement_id", None),
            "balise": getattr(self, "balise", None) or "autre",
            "mode_production": getattr(self, "mode_production", None) or "fixe",
        }

File: frontend/src/utils/gainPassif.js
/**
 * Variation nette de stock au prochain tour pour une ressource,
 * en appliquant les règles actives dans l’ordre des id (aligné backend).
 */
export function deltaNetProchainTour(stockQuantite, gainsPourRessource) {
  const list = (gainsPourRessource || [])
    .filter((g) => g.actif && Number(g.delai_tours ?? 0) <= 0)
    .sort((a, b) => (a.id ?? 0) - (b.id ?? 0));
  // Pourcentage = % de la production du tour (pas du stock).
  let prod = 0;
  let total = 0;
  for (const g of list) {
    const mode = g.mode_production || "fixe";
    const d =
      mode === "pourcentage"
        ? Math.trunc((prod * g.quantite_par_tour) / 100)
        : g.quantite_par_tour;
    prod += d;
    total += d;
  }
  return total;
}

export const BALISE_LABELS = {
  science: "Science",
  politique: "Politique",
  evenement: "Événement",
  batiment: "Bâtiment",
  recolte_fructueuse: "Récolte fructueuse",
  autre: "Autre",
};

export function libelleBalise(b) {
  return BALISE_LABELS[b] ?? BALISE_LABELS.autre;
}

export function formatEffetProduction(g) {
  const mode = g.mode_production || "fixe";
  const q = g.quantite_par_tour;
  if (mode === "pourcentage") {
    const s = q > 0 ? `+${q}` : String(q);
    return `${s} % de la prod. du tour (avant cette ligne)`;
  }
  const s = q > 0 ? `+${q}` : String(q);
  return `${s} unités`;
}
